sources/appendix-value-extraction.asciidoc

```diff
@@ -0,0 +1,437 @@
+[source,java]
+----
+private Tuple<@NotNull @Email String, @NotNull @Min(1) Integer> tuple;
+----
+
+==== Invalid examples
+
+* Extractor types must declare exactly one parameter as the extracted value. Hence this extractor type is invalid:
+
+[source,java]
+----
+public class DoubleExtractor implements ValueExtractor<Multimap<@ExtractedValue ?, @ExtractedValue ?>> { ... }
+----
+
+* No most specific extractor can be found unambiguously:
+
+[source,java]
+----
+public interface CachedValue<V> {
+    V getCachedValue();
+}
+----
+
+[source,java]
+----
+public interface RealValue<V> {
+    V getRealValue();
+}
+----
+
+[source,java]
+----
+public class CachableValue<V> implements CachedValue<V>, RealValue<V> { ... }
+----
+
+[source,java]
+----
+public class CachedValueExtractor implements ValueExtractor<CachedValue<@ExtractedValue ?>> { ... }
+----
+
+[source,java]
+----
+public class RealValueExtractor implements ValueExtractor<RealValue<@ExtractedValue ?>> { ... }
+----
+
+[source,java]
+----
+private CachableValue<@Min(1) Integer> foo;
+----
+
+Validation of `foo` will fail, as none of the two matching extractors is more specific than the other one. An extractor for `CachableValue` must be added, resolving the ambiguity.
+
+==== XML based configuration
+
+TODO
+
+==== Metadata retrieval
+
+TODO
+
+==== Misc.
+
+* Regarding group sequences and default group sequences, the same rules apply for type-level constraints as they apply for regular element-level constraints.
+
+=== Open questions
+
+* `ConstraintsApplyTo` only allows one behavior per annotated element. Should it be per constraint? E.g. for `@NotNull @Email StringProperty email` it may be desirable to apply `@NotNull` to the wrapper but `@Email` to the wrapped value. That's not possible currently.
+* Should type-level constraints be validated when the container is null?
+* What to return from `PropertyDescriptor#getElementClass()` if there is a field of type `Foo` but a getter of type `Optional<Foo>`. So far, BV assumed the types of field and getter to be the same and exposed a single property descriptor (which btw. also may fall apart as of BV 1.1 when the field is of a sub-type of the getter's type). What to return here?
+* Should the presence of type-level constraints alone trigger cascaded validation? It doesn't seem right, but it may be useful when e.g. considering the case of `Tuple` above.
+* Should we allow extractors to be defined for specific parameterizations of generic types, e.g.
+
+[source,java]
+----
+public class ListOfIntegerExtractor implements ValueExtractor<List<@ExtractedValue Integer>> { ... }
+
+public class ListOfStringExtractor implements ValueExtractor<List<@ExtractedValue String>> { ... }
+----
+
+I can't see a compelling use case for this (when would extractor behavior differ between different parameterizations of the same generic type) and am leaning towards only supporting the wildcard parameterization (`implements ValueExtractor<List<@ExtractedValue ?>>`).
```

Comment:
I'd say +1 for only supporting the wildcard for now. It looks like the algorithm could be refined if we find a usage later.